File: typescript/src/rules/no-ungoverned-tool-call.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 MuVeraAI Corporation

/**
 * Rule: no-ungoverned-tool-call
 *
 * Detects tool invocations (tool.run(), tool.execute(), tool.invoke(), etc.)
 * that are not preceded by a governance check (engine.check(),
 * governance.check(), trust.verify(), etc.) within the same function scope.
 *
 * Rationale: Every tool action an agent takes must be authorised by the
 * governance layer before execution. Skipping the check means the agent
 * can take arbitrary actions without policy enforcement.
 */

import type { Rule } from 'eslint';
import type { CallExpression, Node } from 'estree';
import {
  isToolCall,
  isGovernanceCheck,
  findEnclosingFunction,
  collectCallExpressions,
  anyCallPrecedesNode,
} from '../utils.js';

const rule: Rule.RuleModule = {
  meta: {
    type: 'problem',
    docs: {
      description:
        'Require a governance check before every tool invocation. Tool calls without a prior engine.check() / governance.check() in the same scope are ungoverned and violate agent policy.',
      url: 'https://github.com/muveraai/aumos-oss/blob/main/governance-linter/docs/rules.md#no-ungoverned-tool-call',
    },
    messages: {
      missingGovernanceCheck:
        "'{{callExpr}}' is a tool invocation but no governance check (e.g. engine.check() or governance.check()) was found before it in the enclosing function. Add a check to authorise this action.",
    },
    schema: [
      {
        type: 'object',
        properties: {
          additionalToolPatterns: {
            type: 'array',
            items: { type: 'string' },
            description: 'Additional object names to treat as tool objects.',
          },
          additionalCheckPatterns: {
            type: 'array',
            items: { type: 'string' },
            description: 'Additional method names to treat as governance checks.',
          },
        },
        additionalProperties: false,
      },
    ],
  },

  create(context: Rule.RuleContext): Rule.RuleListener {
    return {
      CallExpression(node: Node): void {
        const callNode = node as CallExpression;

        if (!isToolCall(callNode)) return;

        // Build a human-readable callee string for the error message
        const callExpr = context.getSourceCode().getText(callNode.callee);

        // Walk up the AST to find the enclosing function body
        const ancestors = context.getAncestors();
        const enclosingFunction = findEnclosingFunction(callNode, ancestors);

        if (!enclosingFunction) {
          // Top-level tool call — no enclosing function, must still be governed
          context.report({
            node,
            messageId: 'missingGovernanceCheck',
            data: { callExpr },
          });
          return;
        }

        // Collect all call expressions inside the enclosing function
        const allCalls = collectCallExpressions(enclosingFunction);

        // Check if any governance check precedes this tool call
        const hasGovernanceCheck = anyCallPrecedesNode(
          allCalls,
          callNode,
          isGovernanceCheck,
        );

        if (!hasGovernanceCheck) {
          context.report({
            node,
            messageId: 'missingGovernanceCheck',
            data: { callExpr },
          });
        }
      },
    };
  },
};

export default rule;

File: python/src/governance_linter/rules/no_ungoverned_tool_call.py
# SPDX-License-Identifier: Apache-2.0
# Copyright (c) 2026 MuVeraAI Corporation

"""
Rule: no-ungoverned-tool-call

Detects tool invocations (tool.run(), tool.execute(), tool.invoke(), etc.)
that are not preceded by a governance check (engine.check(),
governance.check(), trust.verify(), etc.) within the same function scope.

Rationale: Every action an agent takes through a tool must be authorised by
the governance layer before execution. Skipping the check means the agent can
perform arbitrary actions without policy enforcement.
"""

import ast

from .base import BaseRule

# Object names treated as tool handles
TOOL_OBJECTS: set[str] = {"tool", "tools", "agent", "executor"}

# Method names treated as tool invocation verbs
TOOL_METHODS: set[str] = {"run", "execute", "invoke", "call", "dispatch"}

# Object names that carry a governance / trust check
GOVERNANCE_OBJECTS: set[str] = {"engine", "governance", "trust", "policy", "aumos"}

# Method names that constitute a governance check
GOVERNANCE_METHODS: set[str] = {
    "check",
    "verify",
    "validate",
    "authorize",
    "permit",
}


class NoUngovernedToolCall(BaseRule):
    """Flag tool invocations that lack a prior governance check in the same scope."""

    rule_id = "no-ungoverned-tool-call"
    description = (
        "Require a governance check before every tool invocation. Tool calls "
        "without a prior engine.check() / governance.check() in the same scope "
        "are ungoverned and violate agent policy."
    )

    def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
        self._check_function_body(node)
        self.generic_visit(node)

    def visit_AsyncFunctionDef(self, node: ast.AsyncFunctionDef) -> None:
        self._check_function_body(node)
        self.generic_visit(node)

    def _check_function_body(
        self, func_node: ast.FunctionDef | ast.AsyncFunctionDef
    ) -> None:
        """Walk all calls in this function and flag ungoverned tool calls."""
        all_calls = self.collect_calls(func_node)

        for call in all_calls:
            if not self.is_method_call(call, objects=TOOL_OBJECTS, methods=TOOL_METHODS):
                continue

            target_line: int = getattr(call, "lineno", 0)

            has_prior_check = self.calls_before(
                all_calls,
                target_line,
                objects=GOVERNANCE_OBJECTS,
                methods=GOVERNANCE_METHODS,
            )

            if not has_prior_check:
                callee_text = self._format_callee(call)
                self.report(
                    call,
                    f"'{callee_text}' is a tool invocation but no governance check "
                    "(e.g. engine.check() or governance.check()) was found before it "
                    "in the enclosing function. Add a check to authorise this action.",
                )

    @staticmethod
    def _format_callee(call: ast.Call) -> str:
        if isinstance(call.func, ast.Attribute) and isinstance(call.func.value, ast.Name):
            return f"{call.func.value.id}.{call.func.attr}"
        return "<unknown>"
